### PR TITLE
Fix toggling favorite and display favorite items in the front of inventory UI

### PR DESCRIPTION
**Pull request by Qrox**
#### Summary
SUMMARY: Bugfixes "Fix toggling favorite and display favorite items in the front of inventory UI"

#### Purpose of change
Toggling favorite in inventory UI wasn't working. This PR fixes it. Also, make favorite items displayed in the front of each category, since they are favorites after all.

#### Describe the solution
Assign favorite flags to selected items directly, instead of going the roundabout way of first finding the items in inventory/on the map/in vehicle, which are not working anymore after the nested container change.

Move the tests for invlets out of `inventory_selector_preset::sort_compare` and into the sorting lambda, so lettered items are always put in the front in every inventory menu. Added a second test for favorite flag, so favorite items are also put in the front.

Also fixed the sorting code of inventory menu so it now only sorts once before displaying the menu. This allows the items to keep their position when favorited, until the menu is reopened, so you can batch favorite items without the items jumping around.

In addition, consuming menu now moves the cursor to the previously selected item if possible, so when the items are re-sorted the cursor still points to the last consumed item.

Also removed a hack used to refresh inventory UI when favoriting items, and a bug where cursor disappaers when two columns are merged due to low screen space.

#### Testing
Toggling favorite is working now. Favorited items are now displayed in the front of each category when opening the menu. Consumed a few items after favoriting and the cursor still points to the original item before consuming.

### CODE REVIEW

**Comment by KorGgenT** [comment]
image
something went wrong when i tried to favorite this stack of 9 processor boards. it only favorited a couple, then i pressed * again and it favorited some more, now it's stacked weird in the ui. this could also be a problem with the stacking code and not anything to do with this PR however.

**Comment by Qrox** [comment]
Well, I can't reproduce that, so there might be some more specific condition for that to happen.

**Comment by KorGgenT** [comment]
[Vanilla.zip](https://github.com/CleverRaven/Cataclysm-DDA/files/4943376/Vanilla.zip)
this is the save with that behavior, if you can't reproduce it here i guess my save is just being weird

**Comment by Qrox** [comment]
I favorited the stack of 9 processor boards and they were correctly marked as favorite. However the stack was re-sorted to display at the bottom of the category, and the place where it was original displayed now displays another stack of 2 processor boards. Probably that's why you had the impression that only part of the stack was marked as favorite?

**Comment by Qrox** [comment]
@KorGgenT I made it so that favorited items are not re-sorted until the menu is reopened, which should fix the behavior you encountered.

File: src/game_inventory.cpp [hunk 1]
```diff
@@ -131,26 +131,13 @@ static item_location inv_internal( Character &u, const inventory_selector_preset
...
-    if( u.has_activity( consuming ) && u.activity.values.size() >= 2 ) {
-        init_pair.first = u.activity.values[0];
-        init_pair.second = u.activity.values[1];
-        init_selection = true;
-    }
-    if( u.has_activity( consuming ) && !u.activity.str_values.empty() ) {
-        init_filter = u.activity.str_values[0];
-        has_init_filter = true;
-    }
+    bool init = true;
```
Comment (ZhilkinSerg) [inline src/game_inventory.cpp:140]:
```suggestion
```

File: src/game_inventory.cpp [hunk 2]
```diff
@@ -159,15 +146,20 @@ static item_location inv_internal( Character &u, const inventory_selector_preset
...
+            // Set position after filter to keep cursor at the right position
+            bool position_set = false;
+            if( !u.activity.targets.empty() ) {
+                position_set = inv_s.select_one_of( u.activity.targets );
+            }
+            if( !position_set && u.activity.values.size() >= 2 ) {
+                inv_s.select_position( std::make_pair( u.activity.values[0], u.activity.values[1] ) );
+            }
         }
+        init = false;
```
Comment (ZhilkinSerg) [inline src/game_inventory.cpp:162]:
```suggestion
```

File: src/game_inventory.cpp (hunk 1 above)
Comment (Qrox) [reply src/game_inventory.cpp:140]: Why? `init` is used at line 149.

Comment (ZhilkinSerg) [reply src/game_inventory.cpp:140]: Don't mind it. I forgot to remove that suggestion - it was added by mistake.

**Pull request merged**